### ISSUE TITLE
assistance deletion of key/value

### ISSUE THREAD

**Issue by Phil-ThePower-Pearce** [russellballestrini/nested-lookup#35]
Hi,

Im using this, I have a dictionary I recieve from a shodan api. When I list the keys I get all that I expect..

keys = get_all_keys(shodan_data)
```
['region_code', 'ip', 'postal_code', 'country_code', 'city', 'dma_code', 'last_update', 'latitude', 'tags', 'area_code', 'country_name', 'hostnames', 'org', 'data', '_shodan', 'id', 'options', 'ptr', 'module', 'crawler', 'product', 'hash', 'os', 'opts', 'vulns', 'heartbleed', 'ip', 'isp', 'http', 'html_hash', 'robots_hash', 'redirects', 'securitytxt', 'title', 'sitemap_hash', 'robots', 'favicon', 'hash', 'data', 'location', 'host', 'html', 'location', 'components', 'React', 'categories', 'webpack', 'categories', 'Stripe', 'categories', 'Gatsby', 'categories', 'Google Font API', 'categories', 'server', 'sitemap', 'securitytxt_hash', 'cpe', 'port', 'ssl', 'dhparams', 'tlsext', 'id', 'name', 'id', 'name', 'id', 'name', 'versions', 'acceptable_cas', 'cert', 'sig_alg', 'issued', 'expires', 'expired', 'version', 'extensions', 'critical', 'data', 'name', 'data', 'name', 'critical', 'data', 'name', 'data', 'name', 'data', 'name', 'data', 'name', 'data', 'name', 'data', 'name', 'data', 'name', 'fingerprint', 'sha256', 'sha1', 'serial', 'subject', 'CN', 'pubkey', 'type', 'bits', 'issuer', 'C', 'CN', 'O', 'cipher', 'version', 'bits', 'name', 'chain', 'alpn', 'hostnames', 'location', 'city', 'region_code', 'area_code', 'longitude', 'country_code3', 'country_name', 'postal_code', 'dma_code', 'country_code', 'latitude', 'timestamp', 'domains', 'org', 'data', 'asn', 'transport', 'ip_str']
```
when i use
results = nested_delete(shodan_data, "html")
the key is not deleted, am I using this correctly or making a massive rookie mistake?
If someone could point me in the right direction that would be great?

**Comment by russellballestrini** [comment]
Could you please post a complete example `shodan_data` so I may try to reproduce and write a unit test case?

**Comment by Phil-ThePower-Pearce** [comment]
[https://pastebin.com/18AkLzmP](url)
Its an mix of amass /shodan results

**Comment by russellballestrini** [comment]
Any chance you could help me pair down this failing test case into one which is much smaller?

**Comment by Phil-ThePower-Pearce** [comment]
https://pastebin.com/40y90Tnu
This is just one record, that better?

**Comment by russellballestrini** [comment]
Yes, thank you. 

Fir the record the key is not deleted from the newly returned document assigned to `results`?

I'll try to reproduce shortly.

**Comment by russellballestrini** [comment]
I'm not able to load that pastebin with Python's JSON. Do you happen to have a copy of this data in Python form? (lists and dicts)

```
>>> import json
>>> d = json.loads(d_json)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
    d = json.loads(d_json)
  File "/usr/lib64/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python2.7/json/decoder.py", line 380, in raw_decode
    obj, end = self.scan_once(s, idx)
ValueError: Invalid \escape: line 79 column 31 (char 2027)
```

**Comment by txjacob** [comment]
I've been having this issue this morning too. I've got a value in my dictionary that contains a bool, and it seems to trip up the library if the bool contains 'false'. I seem to have traced it to:
`    elif isinstance(document, dict):
        if document.get(key):
            del document[key]
        for dict_key, dict_value in iteritems(document):
            _nested_delete(document=dict_value, key=key)`

Specifically `if document.get(key):` which looks like if the value contains false, it will make it look like the value is not in the dictionary.

**Comment by txjacob** [comment]
I made a pull request for what fixed the issue for me.
See #36

**Comment by Phil-ThePower-Pearce** [comment]
```
`[
  {
    "name": "pm.appsec.eu",
    "ip": "85.214.148.181",
    "desc": "STRATO STRATO AG",
    "cidr": "85.214.0.0/15",
    "asn": 6724,
    "tag": "cert",
    "source": "Crtsh",
    "shodan_present": true,
    "shodan_data": {
      "region_code": null,
      "ip": 1440126133,
      "postal_code": null,
      "country_code": "DE",
      "city": null,
      "dma_code": null,
      "last_update": "2020-06-10T08:39:17.791127",
      "latitude": 51.2993,
      "tags": [
        "self-signed"
      ],
      "area_code": null,
      "country_name": "Germany",
      "hostnames": [
        "h2807224.stratoserver.net"
      ],
      "org": "Strato AG",
      "data": [
        {
          "_shodan": {
            "id": "05f933c9-7be0-46e8-9b01-fb5d4d12055b",
            "options": {},
            "ptr": true,
            "module": "rdp",
            "crawler": "70752434fdf0dcec35df6ae02b9703eaae035f7d"
          },
          "hash": -1660128709,
          "os": null,
          "tags": [
            "self-signed"
          ],
          "opts": {
            "raw": "030000130ed00000123400021f080002000000"
          },
          "ip": 1440126133,
          "isp": "Strato AG",
          "port": 3389,
          "ssl": {
            "dhparams": {
              "prime": "",
              "public_key": "",
              "bits": 2048,
              "generator": ""
            },
            "tlsext": [
              {
                "id": 65281,
                "name": "renegotiation_info"
              }
            ],
            "versions": [
              "TLSv1",
              "-SSLv2",
              "-SSLv3",
              "TLSv1.1",
              "TLSv1.2"
            ],
            "acceptable_cas": [],
            "cert": {
              "sig_alg": "sha256WithRSAEncryption",
              "issued": "20200415175907Z",
              "expires": "20201015175907Z",
              "expired": false,
              "version": 2,
              "extensions": [
                {
                  "data": "0\\n\\x06\\x08+\\x06\\x01\\x05\\x05\\x07\\x03\\x01",
                  "name": "extendedKeyUsage"
                },
                {
                  "data": "\\x03\\x02\\x040",
                  "name": "keyUsage"
                }
              ],
              "fingerprint": {
                "sha256": "25aa48c21998ea6fa46a32b096e02742804bc39b67794dc1db7bf6ad10e43b8b",
                "sha1": "cd581e62b7ab7d8bf9257bccd896723aacfdd422"
              },
              "serial": 102075361888227116599151106298126556257,
              "subject": {
                "CN": "h2807224"
              },
              "pubkey": {
                "type": "rsa",
                "bits": 2048
              },
              "issuer": {
                "CN": "h2807224"
              }
            },
            "cipher": {
              "version": "TLSv1/SSLv3",
              "bits": 256,
              "name": "ECDHE-RSA-AES256-GCM-SHA384"
            },
            "chain": [
              ""
            ],
            "alpn": []
          },
          "hostnames": [
            "h2807224.stratoserver.net"
          ],
          "location": {
            "city": null,
            "region_code": null,
            "area_code": null,
            "longitude": 9.491,
            "country_code3": null,
            "country_name": "Germany",
            "postal_code": null,
            "dma_code": null,
            "country_code": "DE",
            "latitude": 51.2993
          },
          "timestamp": "2020-06-10T08:39:17.791127",
          "domains": [
            "stratoserver.net"
          ],
          "org": "Strato AG",
          "data": "Remote Desktop Protocol\n\\x03\\x00\\x00\\x13\\x0e\\xd0\\x00\\x00\\x124\\x00\\x02\\x1f\\x08\\x00\\x02\\x00\\x00\\x00",
          "asn": "AS6724",
          "transport": "tcp",
          "ip_str": "85.214.148.181"
        },
        {
          "_shodan": {
            "id": "51edec54-accf-49f5-bf47-d212add5ac7e",
            "options": {},
            "ptr": true,
            "module": "http",
            "crawler": "23edd73099dd1e26ac30e646b7bb7857d248d542"
          },
          "product": "Microsoft IIS httpd",
          "hash": -1632652600,
          "version": "10.0",
          "opts": {},
          "ip": 1440126133,
          "isp": "Strato AG",
          "http": {
            "html_hash": -2076671977,
            "robots_hash": null,
            "redirects": [],
            "securitytxt": null,
            "title": "STRATO",
            "sitemap_hash": null,
            "robots": null,
            "favicon": null,
            "host": "85.214.148.181",
            "html": "<html>\r\n<head>\r\n<title>STRATO</title>\r\n<meta name=\"author\" content=\"Cipolletta\">\r\n<meta name=\"generator\" content=\"HTML EDITOR\">\r\n</head>\r\n<body bgcolor=\"white\">\r\n<p>&nbsp;</p>\r\n<p align=\"center\">\r\n    <img src=\"setup.jpg\" border=\"0\" alt=\"This web site is now reserved. As of now, no content has been uploaded\" />\r\n</p>\r\n</body>\r\n</html>\r\n",
            "location": "/",
            "components": {},
            "server": "Microsoft-IIS/10.0",
            "sitemap": null,
            "securitytxt_hash": null
          },
          "cpe": [
            "cpe:/a:microsoft:internet_information_server:10.0",
            "cpe:/o:microsoft:windows"
          ],
          "port": 80,
          "os": "Windows Server 2008",
          "hostnames": [
            "h2807224.stratoserver.net"
          ],
          "link": "Ethernet or modem",
          "location": {
            "city": null,
            "region_code": null,
            "area_code": null,
            "longitude": 9.491,
            "country_code3": null,
            "country_name": "Germany",
            "postal_code": null,
            "dma_code": null,
            "country_code": "DE",
            "latitude": 51.2993
          },
          "timestamp": "2020-06-10T03:51:47.317989",
          "domains": [
            "stratoserver.net"
          ],
          "org": "Strato AG",
          "data": "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\nLast-Modified: Wed, 11 Jul 2007 09:05:10 GMT\r\nAccept-Ranges: bytes\r\nETag: \"30ac6b959ac3c71:0\"\r\nServer: Microsoft-IIS/10.0\r\nDate: Wed, 10 Jun 2020 03:51:46 GMT\r\nContent-Length: 338\r\n\r\n",
          "asn": "AS6724",
          "transport": "tcp",
          "ip_str": "85.214.148.181"
        },
        {
          "_shodan": {
            "id": "2a1b96ea-a0c4-470e-b9d0-f6a9203e1939",
            "options": {},
            "ptr": true,
            "module": "smb",
            "crawler": "59f6eda5a6b574c5c0478ef4da1560159fca2427"
          },
          "hash": -2121476472,
          "os": "Windows Server 2016 Standard 14393",
          "smb": {
            "smb_version": 1,
            "os": "Windows Server 2016 Standard 14393",
            "capabilities": [
              "unicode",
              "large-files",
              "nt-smb",
              "rpc-remote-api",
              "nt-status",
              "level2-oplocks",
              "lock-and-read",
              "nt-find",
              "infolevel-passthru",
              "large-readx",
              "large-writex",
              "lwio",
              "extended-security"
            ],
            "software": "Windows Server 2016 Standard 6.3"
          },
          "opts": {},
          "ip": 1440126133,
          "isp": "Strato AG",
          "port": 445,
          "hostnames": [
            "h2807224.stratoserver.net"
          ],
          "location": {
            "city": null,
            "region_code": null,
            "area_code": null,
            "longitude": 9.491,
            "country_code3": null,
            "country_name": "Germany",
            "postal_code": null,
            "dma_code": null,
            "country_code": "DE",
            "latitude": 51.2993
          },
          "timestamp": "2020-06-03T03:07:22.717987",
          "domains": [
            "stratoserver.net"
          ],
          "org": "Strato AG",
          "data": "SMB Status\nAuthentication: enabled\nSMB Version: 1\nCapabilities: unicode,large-files,nt-smb,rpc-remote-api,nt-status,level2-oplocks,lock-and-read,nt-find,infolevel-passthru,large-readx,large-writex,lwio,extended-security\n",
          "asn": "AS6724",
          "transport": "tcp",
          "ip_str": "85.214.148.181"
        }
      ],
      "asn": "AS6724",
      "isp": "Strato AG",
      "longitude": 9.491,
      "country_code3": null,
      "domains": [
        "stratoserver.net"
      ],
      "ip_str": "85.214.148.181",
      "os": null,
      "ports": [
        80,
        3389,
        445
      ]
    }
  }
]`
```

**Comment by Phil-ThePower-Pearce** [comment]
> I'm not able to load that pastebin with Python's JSON. Do you happen to have a copy of this data in Python form? (lists and dicts)
> 
> ```
> >>> import json
> >>> d = json.loads(d_json)
> Traceback (most recent call last):
>   File "<input>", line 1, in <module>
>     d = json.loads(d_json)
>   File "/usr/lib64/python2.7/json/__init__.py", line 339, in loads
>     return _default_decoder.decode(s)
>   File "/usr/lib64/python2.7/json/decoder.py", line 364, in decode
>     obj, end = self.raw_decode(s, idx=_w(s, 0).end())
>   File "/usr/lib64/python2.7/json/decoder.py", line 380, in raw_decode
>     obj, end = self.scan_once(s, idx)
> ValueError: Invalid \escape: line 79 column 31 (char 2027)
> ```
```
[
  {
    "name": "pm.appsec.eu",
    "ip": "85.214.148.181",
    "desc": "STRATO STRATO AG",
    "cidr": "85.214.0.0/15",
    "asn": 6724,
    "tag": "cert",
    "source": "Crtsh",
    "shodan_present": true,
    "shodan_data": {
      "region_code": null,
      "ip": 1440126133,
      "postal_code": null,
      "country_code": "DE",
      "city": null,
      "dma_code": null,
      "last_update": "2020-06-10T08:39:17.791127",
      "latitude": 51.2993,
      "tags": [
        "self-signed"
      ],
      "area_code": null,
      "country_name": "Germany",
      "hostnames": [
        "h2807224.stratoserver.net"
      ],
      "org": "Strato AG",
      "data": [
        {
          "_shodan": {
            "id": "05f933c9-7be0-46e8-9b01-fb5d4d12055b",
            "options": {},
            "ptr": true,
            "module": "rdp",
            "crawler": "70752434fdf0dcec35df6ae02b9703eaae035f7d"
          },
          "hash": -1660128709,
          "os": null,
          "tags": [
            "self-signed"
          ],
          "opts": {
            "raw": "030000130ed00000123400021f080002000000"
          },
          "ip": 1440126133,
          "isp": "Strato AG",
          "port": 3389,
          "ssl": {
            "dhparams": {
              "prime": "",
              "public_key": "",
              "bits": 2048,
              "generator": ""
            },
            "tlsext": [
              {
                "id": 65281,
                "name": "renegotiation_info"
              }
            ],
            "versions": [
              "TLSv1",
              "-SSLv2",
              "-SSLv3",
              "TLSv1.1",
              "TLSv1.2"
            ],
            "acceptable_cas": [],
            "cert": {
              "sig_alg": "sha256WithRSAEncryption",
              "issued": "20200415175907Z",
              "expires": "20201015175907Z",
              "expired": false,
              "version": 2,
              "extensions": [
                {
                  "data": "0\\n\\x06\\x08+\\x06\\x01\\x05\\x05\\x07\\x03\\x01",
                  "name": "extendedKeyUsage"
                },
                {
                  "data": "\\x03\\x02\\x040",
                  "name": "keyUsage"
                }
              ],
              "fingerprint": {
                "sha256": "25aa48c21998ea6fa46a32b096e02742804bc39b67794dc1db7bf6ad10e43b8b",
                "sha1": "cd581e62b7ab7d8bf9257bccd896723aacfdd422"
              },
              "serial": 102075361888227116599151106298126556257,
              "subject": {
                "CN": "h2807224"
              },
              "pubkey": {
                "type": "rsa",
                "bits": 2048
              },
              "issuer": {
                "CN": "h2807224"
              }
            },
            "cipher": {
              "version": "TLSv1/SSLv3",
              "bits": 256,
              "name": "ECDHE-RSA-AES256-GCM-SHA384"
            },
            "chain": [
              ""
            ],
            "alpn": []
          },
          "hostnames": [
            "h2807224.stratoserver.net"
          ],
          "location": {
            "city": null,
            "region_code": null,
            "area_code": null,
            "longitude": 9.491,
            "country_code3": null,
            "country_name": "Germany",
            "postal_code": null,
            "dma_code": null,
            "country_code": "DE",
            "latitude": 51.2993
          },
          "timestamp": "2020-06-10T08:39:17.791127",
          "domains": [
            "stratoserver.net"
          ],
          "org": "Strato AG",
          "data": "Remote Desktop Protocol\n\\x03\\x00\\x00\\x13\\x0e\\xd0\\x00\\x00\\x124\\x00\\x02\\x1f\\x08\\x00\\x02\\x00\\x00\\x00",
          "asn": "AS6724",
          "transport": "tcp",
          "ip_str": "85.214.148.181"
        },
        {
          "_shodan": {
            "id": "51edec54-accf-49f5-bf47-d212add5ac7e",
            "options": {},
            "ptr": true,
            "module": "http",
            "crawler": "23edd73099dd1e26ac30e646b7bb7857d248d542"
          },
          "product": "Microsoft IIS httpd",
          "hash": -1632652600,
          "version": "10.0",
          "opts": {},
          "ip": 1440126133,
          "isp": "Strato AG",
          "http": {
            "html_hash": -2076671977,
            "robots_hash": null,
            "redirects": [],
            "securitytxt": null,
            "title": "STRATO",
            "sitemap_hash": null,
            "robots": null,
            "favicon": null,
            "host": "85.214.148.181",
            "html": "<html>\r\n<head>\r\n<title>STRATO</title>\r\n<meta name=\"author\" content=\"Cipolletta\">\r\n<meta name=\"generator\" content=\"HTML EDITOR\">\r\n</head>\r\n<body bgcolor=\"white\">\r\n<p>&nbsp;</p>\r\n<p align=\"center\">\r\n    <img src=\"setup.jpg\" border=\"0\" alt=\"This web site is now reserved. As of now, no content has been uploaded\" />\r\n</p>\r\n</body>\r\n</html>\r\n",
            "location": "/",
            "components": {},
            "server": "Microsoft-IIS/10.0",
            "sitemap": null,
            "securitytxt_hash": null
          },
          "cpe": [
            "cpe:/a:microsoft:internet_information_server:10.0",
            "cpe:/o:microsoft:windows"
          ],
          "port": 80,
          "os": "Windows Server 2008",
          "hostnames": [
            "h2807224.stratoserver.net"
          ],
          "link": "Ethernet or modem",
          "location": {
            "city": null,
            "region_code": null,
            "area_code": null,
            "longitude": 9.491,
            "country_code3": null,
            "country_name": "Germany",
            "postal_code": null,
            "dma_code": null,
            "country_code": "DE",
            "latitude": 51.2993
          },
          "timestamp": "2020-06-10T03:51:47.317989",
          "domains": [
            "stratoserver.net"
          ],
          "org": "Strato AG",
          "data": "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\nLast-Modified: Wed, 11 Jul 2007 09:05:10 GMT\r\nAccept-Ranges: bytes\r\nETag: \"30ac6b959ac3c71:0\"\r\nServer: Microsoft-IIS/10.0\r\nDate: Wed, 10 Jun 2020 03:51:46 GMT\r\nContent-Length: 338\r\n\r\n",
          "asn": "AS6724",
          "transport": "tcp",
          "ip_str": "85.214.148.181"
        },
        {
          "_shodan": {
            "id": "2a1b96ea-a0c4-470e-b9d0-f6a9203e1939",
            "options": {},
            "ptr": true,
            "module": "smb",
            "crawler": "59f6eda5a6b574c5c0478ef4da1560159fca2427"
          },
          "hash": -2121476472,
          "os": "Windows Server 2016 Standard 14393",
          "smb": {
            "smb_version": 1,
            "os": "Windows Server 2016 Standard 14393",
            "capabilities": [
              "unicode",
              "large-files",
              "nt-smb",
              "rpc-remote-api",
              "nt-status",
              "level2-oplocks",
              "lock-and-read",
              "nt-find",
              "infolevel-passthru",
              "large-readx",
              "large-writex",
              "lwio",
              "extended-security"
            ],
            "software": "Windows Server 2016 Standard 6.3"
          },
          "opts": {},
          "ip": 1440126133,
          "isp": "Strato AG",
          "port": 445,
          "hostnames": [
            "h2807224.stratoserver.net"
          ],
          "location": {
            "city": null,
            "region_code": null,
            "area_code": null,
            "longitude": 9.491,
            "country_code3": null,
            "country_name": "Germany",
            "postal_code": null,
            "dma_code": null,
            "country_code": "DE",
            "latitude": 51.2993
          },
          "timestamp": "2020-06-03T03:07:22.717987",
          "domains": [
            "stratoserver.net"
          ],
          "org": "Strato AG",
          "data": "SMB Status\nAuthentication: enabled\nSMB Version: 1\nCapabilities: unicode,large-files,nt-smb,rpc-remote-api,nt-status,level2-oplocks,lock-and-read,nt-find,infolevel-passthru,large-readx,large-writex,lwio,extended-security\n",
          "asn": "AS6724",
          "transport": "tcp",
          "ip_str": "85.214.148.181"
        }
      ],
      "asn": "AS6724",
      "isp": "Strato AG",
      "longitude": 9.491,
      "country_code3": null,
      "domains": [
        "stratoserver.net"
      ],
      "ip_str": "85.214.148.181",
      "os": null,
      "ports": [
        80,
        3389,
        445
      ]
    }
  }
]
```

**Comment by russellballestrini** [comment]
closed by #36
